scrapy/downloadermiddlewares/retry.py

```diff
@@ -31,6 +31,105 @@
 logger = logging.getLogger(__name__)
 
 
+def get_retry_request(
+    request,
+    *,
+    reason,
+    spider,
+    max_retry_times=None,
+    priority_adjust=None,
+):
+    settings = spider.crawler.settings
+    stats = spider.crawler.stats
+    retry_times = request.meta.get('retry_times', 0) + 1
+    request_max_retry_times = request.meta.get(
+        'max_retry_times',
+        max_retry_times,
+    )
+    if request_max_retry_times is None:
+        request_max_retry_times = settings.getint('RETRY_TIMES')
+    if retry_times <= request_max_retry_times:
+        logger.debug(
+            "Retrying %(request)s (failed %(retry_times)d times): %(reason)s",
+            {'request': request, 'retry_times': retry_times, 'reason': reason},
+            extra={'spider': spider}
+        )
+        new_request = request.copy()
+        new_request.meta['retry_times'] = retry_times
+        new_request.dont_filter = True
+        if priority_adjust is None:
+            priority_adjust = settings.getint('RETRY_PRIORITY_ADJUST')
+        new_request.priority = request.priority + priority_adjust
+
+        if isinstance(reason, Exception):
+            reason = global_object_name(reason.__class__)
+
+        stats.inc_value('retry/count')
+        stats.inc_value(f'retry/reason_count/{reason}')
+        return new_request
+    else:
+        stats.inc_value('retry/max_reached')
+        logger.error("Gave up retrying %(request)s (failed %(retry_times)d times): %(reason)s",
+                        {'request': request, 'retry_times': retry_times, 'reason': reason},
+                        extra={'spider': spider})
+        return None
+
+
+def retry_request(
+    request,
+    *,
+    reason,
+    spider,
+    max_retry_times=None,
+    priority_adjust=None,
+):
+    new_request = get_retry_request(
+        request,
+        reason=reason,
+        spider=spider,
+        max_retry_times=max_retry_times,
+        priority_adjust=priority_adjust,
+    )
+    if new_request:
```

Comment:
not sure if this is necessary, as we can just do this by yielding the result of `get_retry_request` (first approach in the description). I would prefer to have a good example on the docs over providing this function.